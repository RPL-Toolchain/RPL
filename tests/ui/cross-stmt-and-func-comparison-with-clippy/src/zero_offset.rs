//@revisions: inline normal
//@[normal] compile-flags: -Zinline-mir=false

/// Base cases
#[cfg_attr(test, test)]
fn base_case() {
    unsafe {
        let mut m = ();
        let m = &raw mut m;

        let n = m.offset(0);
        //~^ zst_offset
        dbg!(n);

        let n = m.wrapping_add(0);
        //~^ zst_offset
        dbg!(n);
    }
}

/// Cross-function cases
#[cfg_attr(test, test)]
fn cross_function() {
    unsafe fn offset<T>(m: *mut T, n: isize) -> *mut T {
        unsafe { m.offset(n) }
    }
    fn wrapping_add<T>(m: *mut T, n: usize) -> *mut T {
        m.wrapping_add(n)
        //~[inline]^ unchecked_pointer_offset
    }
    unsafe {
        let mut m = ();
        let m = &raw mut m;

        let n = offset(m, 0);
        //~[inline]^ zst_offset
        dbg!(n);

        let n = wrapping_add(m, 0);
        //~[inline]^ zst_offset
        dbg!(n);
    }
}

pub(crate) fn main() {
    base_case();
    cross_function();
}
