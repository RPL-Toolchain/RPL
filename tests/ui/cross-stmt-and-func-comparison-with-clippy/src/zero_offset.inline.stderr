error: offset calculation on zero-sized value
  --> tests/ui/cross-stmt-and-func-comparison-with-clippy/src/zero_offset.rs:11:19
   |
LL |         let n = m.offset(0);
   |                   ^^^^^^^^^
   |
   = note: `()` is a zero-sized type, offsetting a pointer to it will always yield the same address
   = note: `#[deny(rpl::zst_offset)]` on by default

error: offset calculation on zero-sized value
  --> tests/ui/cross-stmt-and-func-comparison-with-clippy/src/zero_offset.rs:15:19
   |
LL |         let n = m.wrapping_add(0);
   |                   ^^^^^^^^^^^^^^^
   |
   = note: `()` is a zero-sized type, offsetting a pointer to it will always yield the same address

error: offset calculation on zero-sized value
  --> tests/ui/cross-stmt-and-func-comparison-with-clippy/src/zero_offset.rs:35:17
   |
LL |         let n = offset(m, 0);
   |                 ^^^^^^^^^^^^
   |
   = note: `()` is a zero-sized type, offsetting a pointer to it will always yield the same address

error: offset calculation on zero-sized value
  --> tests/ui/cross-stmt-and-func-comparison-with-clippy/src/zero_offset.rs:39:17
   |
LL |         let n = wrapping_add(m, 0);
   |                 ^^^^^^^^^^^^^^^^^^
   |
   = note: `()` is a zero-sized type, offsetting a pointer to it will always yield the same address

error: it is an undefined behavior to offset a pointer using an unchecked integer
  --> tests/ui/cross-stmt-and-func-comparison-with-clippy/src/zero_offset.rs:28:11
   |
LL |         m.wrapping_add(n)
   |           ^^^^^^^^^^^^^^^
   |           |
   |           offset here
   |           pointer used here
   |
   = help: check whether it's in bound before offsetting
   = note: See the safety section in https://doc.rust-lang.org/std/primitive.pointer.html#method.offset
   = note: `-D rpl::unchecked-pointer-offset` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rpl::unchecked_pointer_offset)]`

error: aborting due to 5 previous errors

